Introduce a new storage specific Env API

anand1976 opened the pull request:
The current Env API encompasses both storage/file operations, as well as OS related operations. Most of the APIs return a Status, which does not have enough metadata about an error, such as whether its retry-able or not, scope (i.e fault domain) of the error etc., that may be required in order to properly handle a storage error. The file APIs also do not provide enough control over the IO SLA, such as timeout, prioritization, hinting about placement and redundancy etc.

This PR separates out the file/storage APIs from Env into a new FileSystem class. The APIs are updated to return an IOStatus with metadata about the error, as well as to take an IOOptions structure as input in order to allow more control over the IO. 

The user can set both ```options.env``` and ```options.file_system``` to specify that RocksDB should use the former for OS related operations and the latter for storage operations. Internally, a ```CompositeEnvWrapper``` has been introduced that inherits from ```Env``` and redirects individual methods to either an ```Env``` implementation or the ```FileSystem``` as appropriate. When options are sanitized during ```DB::Open```, ```options.env``` is replaced with a newly allocated ```CompositeEnvWrapper``` instance if both env and file_system have been specified. This way, the rest of the RocksDB code can continue to function as before.

This PR also ports PosixEnv to the new API by splitting it into two - PosixEnv and PosixFileSystem. PosixEnv is defined as a sub-class of CompositeEnvWrapper, and threading/time functions are overridden with Posix specific implementations in order to avoid an extra level of indirection. 

The ```CompositeEnvWrapper``` translates ```IOStatus``` return code to ```Status```, and sets the severity to ```kSoftError``` if the io_status is retryable. The error handling code in RocksDB can then recover the DB automatically.

siying commented on include/rocksdb/status.h:
```diff
@@ -18,31 +18,16 @@
 namespace rocksdb {
 
-class Status {
+template <class T>
```
RocksDB code base has very limited use of template (which follows the guideline of Google C++ Style https://google.github.io/styleguide/cppguide.html#Template_metaprogramming ), and I don't remember any public interface uses template. This makes the public interface harder to understand. I suggest we think twice before using template here.

anand1976 replied:
Fair enough. I wanted to reuse the code/subcodes and template seemed to be the way to do that. But on further reflection, it allows us to reuse the values but the names have to be qualified by classname. Maybe we can use ```Status::Code``` and ```Status::SubCode``` directly in ```IOStatus```? That way, we can have a single set of error codes, but have a distinct set of named constructors in ```IOStatus``` to indicate what's allowable.

anand1976 commented on include/rocksdb/fs_env.h:
```diff
@@ -0,0 +1,1323 @@
+    std::string file_path;
+
+    // To be set by the FSEnv implementation
+    std::string msg;
```
Yes, I think a map is a good idea to pass back counters that can be used for debugging. We could have a map<string, string> and require the env implementation to format counters as string?

anand1976 commented on include/rocksdb/fs_env.h:
```diff
@@ -0,0 +1,1323 @@
+  static const char* Type() { return "Environment"; }
+
+  // Loads the environment specified by the input value into the result
+  static IOStatus LoadEnv(const std::string& value, FSEnv** result);
```
There is a generic implementation of ```Env::LoadEnv``` that instantiates an Env object from the ObjectRegistry. I guess we need to do the same for ```FSEnv```.

mrambacher replied:
Personally, I think the LoadFSEnv should return a std:::shared_ptr<FSEnv> *.  The use of the static pointers is problematic from a memory management perspective, IMO.

mrambacher commented on include/rocksdb/fs_env.h:
```diff
@@ -0,0 +1,1323 @@
+
+  virtual ~FSEnv();
+
+  static const char* Type() { return "Environment"; }
```
This needs to be something other than "Environment", like "FSEnvironment", to prevent it from clashing with the Env value.

mrambacher commented on include/rocksdb/fs_env.h:
```diff
@@ -0,0 +1,1323 @@
+  // These are hints and are not necessarily guaranteed to be honored.
+  // More hints can be added here in the future to indicate things like
+  // storage media (HDD/SSD) to be used, replication level etc.
+  struct IOOptions {
```
I can see where you might want this on a per-file basis, but why do you want them on a per-request basis?  Can you give an example where one read/write from a file would be a different priority than another to the same file?

mrambacher replied:
The timeout might be a per-operation parameter, but the IOType is not likely to be (are you suggesting the same file might be used for different types of IO?  I also doubt that read or write operations to a single IO file would have different priorities (this read his high but the next one is low does not make sense to me).

anand1976 replied:
We use a single TableReader object for user reads and compaction reads, so yes its possible to have different priority IOs to the same file. Even if we use a separate TableReader for compaction, I wonder if we might want to allow users to specify a priority. All user reads would go through the same file.

mrambacher replied:
Why use IOOptions for this though, rather than using the ReadOptions that are already present?  Why add a new struct with new parameters rather than extending the one that already exists?  It seems like there are an awful lot of places where you would need to update (SST reader/writer, trace methods, etc)

mrambacher commented on include/rocksdb/fs_env.h:
```diff
@@ -0,0 +1,1323 @@
+                                   const IOOptions& io_opts,
+                                   std::unique_ptr<FSWritableFile>* result,
+                                   const EnvOptions& env_opts,
+                                   IODebugContext* dbg) = 0;
```
Should the IOOptions and EnvOptions be merged here?  Why have both arguments?

Also, is there a reason to rename the file classes to FS* names.  It seems like it will cause a lot of changes for not much gain.

mrambacher commented on include/rocksdb/fs_env.h:
```diff
@@ -0,0 +1,1323 @@
+
+  // A structure to pass back some debugging information from the FSEnv
+  // implementation to RocksDB in case of an IO error
+  struct IODebugContext {
```
I am not sure I understand the use of this struct.  Why would there be a different one for every API call?  Who is managing this class (it is usually passed in as a pointer -- can it be null? what is the lifetime, etc?)

Would this be better as something that is a member of the FSEnv?

anand1976 replied:
I envisioned this being similar to PerfContext or IOStatsContext, i.e it could keep debugging counters for a single IO operation. I think its definition will have to go through a couple of iterations before it becomes meaningful. I will update the comments to clarify allocation, lifetime etc.

mrambacher commented on include/rocksdb/io_status.h:
```diff
@@ -0,0 +1,149 @@
+
+namespace rocksdb {
+
+class IOStatus : public BaseStatus<IOStatus> {
```
I understand the need for a new IOStatus class to add all of the extensions to Status.  But would it be better to have IOStatus extend Status and possibly have a new "Code" in Status for IO errors?  Then code could treat the return from a FS operation as either a Status or IOStatus if they cared about the specific sub-codes for retrying the errors.

mrambacher replied:
If an IOStatus extends a Status, then code like:
  Status s = <fs operation> 
and 
  IOStatus s = <fs operation>
could both be valid.  As you have it defined, the Status and IOStatus are not related, and therefore not interchangeable.

anand1976 replied:
Yes, I agree this would reduce the code changes required. I'm concerned that by implementing it this way, it would be very easy for developers to ignore IOStatus. Maintaining the error handling features would be difficult. @siying Any thoughts on this?

mrambacher commented on include/rocksdb/fs_env.h:
```diff
@@ -0,0 +1,1323 @@
+    // To be set by the FSEnv implementation
+    std::string msg;
+
+    IODebugContext(IOOptions& options) { opts = options; }
```
If a IODebugContext is created from an IOOptions, why does every API require both?  Can we do with just one of these?

riversand963 commented on include/rocksdb/io_status.h:
```diff
@@ -0,0 +1,149 @@
+  void SetDataLoss(bool data_loss) { data_loss_ = data_loss; }
+  void SetScope(IOErrorScope scope) { scope_ = scope; }
+
+  bool retryable() { return retryable_; }
```
Nit: `bool retryable() const`

mrambacher commented on include/rocksdb/fs_env.h:
```diff
@@ -0,0 +1,1323 @@
+// of the APIs is of type IOStatus, which can indicate an error code/sub-code,
+// as well as metadata about the error such as its scope and whether its
+// retryable.
+class FSEnv {
```
I agree with that comment.  I was wondering why not call it simply "FileSystem"

anand1976 commented:
@siying @mrambacher @riversand963 I have updated the PR with a port of PosixEnv. I've also incorporated your comments and feedback. The PR description has also been updated with the basic design. PTAL.

anand1976 commented:
The extra wrapper for every file system call is a fair point. I think this can be solved by inlining ```CustomEnvWrapper```. In ```CustomEnvWrapper```, it can fall back to ```Env``` if no ```FileSystem``` has been specified.

anand1976 commented:
@siying Your other point about not knowing how effective is the new interface is more complicated. In this PR, I have focused on basic error recovery for retryable errors. The DB should be able to automatically recover (ignoring the scope of the error) from errors that have ```io_status.retryable()``` set to true (ENOSPC being the only one from Posix as of now). It should work transparently for other such errors.

I don't think we will know the effectiveness of the other aspects of the new interface, such as timeouts and priority, until there is an end to end implementation (including a functioning FileSystem implementation from WS). So I think modifying the DB to specify timeout/priority for IO operations should be deferred to a follow-on PR. In the meantime, it should be made clear that the new interface is still experimental, in order to discourage developers from porting their Env implementations to it right away. Thoughts?

siying commented:
@anand1976 I don't think we can easily inline virtual functions. More importantly, this is not how we want it to eventually look like. Eventually, we want the new API used everywhere, and legacy Env implementation will be wrapped as the new one. So I'm curious how hard it is to skip the reverse wrapped stage and just do what we want to do eventually now.

siying commented:
@anand1976 I totally agree that we should delay the implementation of all the timeout, io priority, etc. I totally support this plan. Even the retryable can be delayed. My comment is not about having those implemented, but about which interface is used. If DB still uses the old Env interface, we do not really see how the new interface is used. I'm not talking about the new options like functions, etc, but the functionalities already in use. I know it is very time consuming to go through all the code base and chase where they are used. But this is needed anyway. By going through it earlier, we can also look at whether the interface is a good one for all those places in longer term, which is helpful to avoid changing the interface again.

anand1976 commented:
The core RocksDB code reads/writes to files through RandomAccessFileReader/WritableFileWriter. It should be relatively straightforward to modify those two to use the new interfaces, as well as deal with the ripple effects (since they would need to be constructed with FSRandomAccessFile/FSWritableFile). This would make it easier in the future to add support in RocksDB for fine-grained error handling, timeouts and IO prioritization.

anand1976 commented:
@mrambacher 

> The new methods take a few additional arguments (IOOptions, IODebugContext). I am not sure what the point is of these arguments -- why not extend EnvOptions for IOOptions? And if you need a debug context, why not make it part of the constructor rather than the individual methods? I do not see what the usage pattern is for this argument yet.

The scope of EnvOptions and IOOptions is different. EnvOptions is for file creation, whereas IOOptions is for individual file operations. For the New*File methods, I agree IOOptions and EnvOptions can be combined (maybe the latter can embed the former). IODebugContext is envisioned to be used for per file request tracing. It cannot be embedded in the file object since the file access is multi-threaded.

> I do not see the need for the FSXXX classes and methods. If the goal is to replace the existing methods, why not just do that rather than create this additional wrapper layer?

Replacing the existing methods would break backward compatibility. There are a number of users with custom Env implementations and requiring changes in them would complicate things. We could overload the methods, but we wouldn't be able to define the new ones as pure virtual (doing so would break compatibility).

> I do not understand the relationship between Env and FileSystem. I would have thought that an Env has a FileSystem in it but this appears not to be the case. Additionally, I am concerned about the lifetime of a FileSystem. It seems like it needs to be a std:shared_ptr, which it is not currently (I think the same should be true of Env BTW).

Agree about FileSystem being a shared_ptr. There's technically no relationship between Env and FileSystem. Going forward, we'd like to have the ability to use multiple FileSystems in the same DB - perhaps for tiering or some other use case.

anand1976 commented:
@siying PTAL. The new commit ports most of the core RocksDB code to use the new API. The RandomAccessFileReader, SequenceFileReader and WritableFileWriter classes call the new API. Tools and tests still use the old API, either directly or through wrapper classes where needed. There may be some small cleanup needed, which I'll do once we agree on the basic structure of the code.

mrambacher commented:
I know I sound like a broken record here, but strongly believe that the additional arguments to the methods is going to be confusing and add little value.  I would much prefer that the IOOptions and EnvOptions have some is-a relationship to one another, rather than passing two in.

As far as the relationship between Env and FileSystem, I still think there is a relationship.  At the moment, a large percentage of the methods of Env are FileSystem related.  I think an Env has three components -- FileSystem, Threads, and Clock/System.  I do not believe that having an Env has-a FileSystem causes any future problems with multiple FileSystems in a single DB.

Finally, I also think that the IODebugContext should be a std::shared_ptr and should ONLY be an argument to the NewXXXFile methods.  I have a hard time envisioning when you would like to debug specific calls to read for specific files and how you would implement that.  I think making it a call to the File constructor could be useful however.  But I also believe it should be a std::shared_ptr to guarantee its lifefime.

I also think that all of the existing Env in the code base (HDFS, Memory, test ones) should be ported to this new API.  That would help validate that the API is correct for more than a single FileSystem implementation.

siying commented:
@mrambacher the main motivation of the separation of Env and FileSystem is that, we plan to support multiple file systems and sharing some other resource like thread pool, timing, etc. So separating file system related functionality feels something we have to do. I agree that there is going to be a painful transition, but we can't think of a better way. Do you have any suggestion to solve this problem?

siying commented on include/rocksdb/file_system.h:
```diff
@@ -0,0 +1,1353 @@
+  // The returned file will only be accessed by one thread at a time.
+  virtual IOStatus NewSequentialFile(const std::string& fname,
+                                     const IOOptions& io_opts,
+                                     const EnvOptions& env_opts,
```
Do we need to put EnvOptions here? Ideally we don't. I know consolidating the two options may be painful, but fow now, can we build a two way converter so that the API can be clean?

anand1976 replied:
The scope of EnvOptions is a file, for the duration of the file being open. The scope of IOOptions is the specific IOs that would happen as part of file open. How about having a FileOptions with IOOptions embedded in it? Other methods such as FSRandomAccessFile::Read() can take IOOptions as an argument, while methods such as FieSystem::NewRandomAccessFile() can take a FileOptions as argument.

siying commented on include/rocksdb/io_status.h:
```diff
@@ -0,0 +1,230 @@
+    IO_ERROR_SCOPE_FILE_SYSTEM,
+    IO_ERROR_SCOPE_FILE,
+    IO_ERROR_SCOPE_RANGE,
+    IO_ERROR_SCOPE_MAX,
```
kIoErrorScopeFileSystem?

mrambacher commented:
@siying  I am not sure how the issues are related.    

If there are multiple FileSystems, there will need to be some mechanism of getting which one you want (like by name).  So Env would probably have two methods:
GetFileSystem() -- returns the "default" one
GetFileSystem(const std::string & id) -- returns the file system based on the input "id".

If you intend to support multiples, something must be the container for storing the multiples, so why not the Env?

I guess one question is which layer knows which FileSystem to use.  At a certain point, someone has to make a decision.

In the places that both an Env and FileSystem are required arguments, what is needed from the Env?  

Another solution would be to invert it.  Instead of an Env having a FileSystem, why not have a FileSystem have an Env?  What from an Env is useful/important to a FileSystem?

siying commented:
@mrambacher I agree that getting a file system from a string may be a good idea. It is useful for many use cases. Where to put it is interesting though. I don't have an opinion on that.
We might also need to support users to provide FileSystem class without a globally name. This is the pattern RocksDB API has been using, block cache, write buffer manager, ThreadPool, etc. These shared resource is represented as C++ objects.

I don't think there are many places where Env and FileSystem need to be passed in together. In some entrance of major functionality, maybe, but I don't expect it will be common.

dhruba commented:
Can we overload the existing Env to encapsulate various combinations of file systems and operating systems.? Every Env is a combination of OperatingSystem + FileSystem + hardware. 

If we can do that elegantly, then we do not have to introduce the FileSystem class to propagate to all the db code.

siying commented:
@dhruba eventually, we will need to allow one DB to store different files to different file systems. The list of file systems to use will belong to DB's configuration, not Env's.

dhruba commented:
@siying agreed. For example, when we had to store files to local xfs as well as AWS S3 fs, we created an AwsEnv. The code that implements this  AwsEnv can then decide where to store which file based on its own criteria. Also, we overloaded EnvOptions so that we can add more config parameters for this decision making.

mrambacher commented:
@dhruba @siying I agree that the system needs to support multiple FileSystems.  But I wonder if those end up being a nested relationship (as is currently implemented) or siblings.  For example, would you expect that you might use a "S3" and an "HDFS" FileSystem for the same DB (with both having a Posix FS as their "default").  How would something like the "Encrypted" FileSystem work in this model?

Perhaps part of the problem is the APIs for opening files are too generic.  Perhaps there should also be different APIs for different types (SST, WAL, etc) files?

siying commented:
@mrambacher good questions. I don't know what other things people wrap Env for. Some of them might get hard to do after we separate FileSystem. Encrypted Env, however, probably won't be a problem in long term. We are redesigning encryption at rest. We are not close to finalize the project but encryption will be a DB feature, so users can request key rotation, etc. The functionality of encryption will be moved out of Env.

mrambacher commented:
@siying @dhruba The question is would one expect a "stack" of FileSystems ("Cloud" has-a "HDFS" has-a "Posix") or an array of them?  If it is an array, how does one FileSystem know which one to hand off requests that it does not handle to?  For example, if you wanted to configure HDFS to be used for higher "level" SST files (say > 3), how does the system know who handles SST files in levels 0-2?  Does HDFS "know" that (via its parent) or is it a separate config?  

What about a FileSystem that might store things remotely (like HDFS or Cloud) but might also want a local "cache".  How does the outer FileSystem know which one is the "cache"?

Without giving it a great deal of thought, it seems like there should be one "FileSystem" per DB, but that a FileSystem can contain other FileSystems in it.  This provides a hierarchy.  And it should be possible to request a FileSystem by name ("Posix", "Cloud", "HDFS") or function ("Cache", "Default").

anand1976 commented:
@mrambacher Stacking of filesystems is something that I expect RocksDB would do, by making decisions about where to store each SST file. For that to work, it has to be an array of filesystems, rather than hidden from RocksDB via a FileSystem hierarchy.

siying commented:
@mrambacher I don't disagree that the approach like an wrapped Env can handle caching may be a good idea. This should still can be done with the new FileSystem interface. There is also value that DB can handle it. For example, can RocksDB put WAL to a different file system than SST files? Can different column families sit on different file systems? Can some cold SST files move to another file system?

The combination Env interface, does solve some problems though. For example, the flexible thread pool and file system combination will be solved, which is good.

dhruba commented:
It feels like Env is a general way to abstract out everything: OS, storage for sst/manifest/log files etc. @mrambacher  whether it is a stack of FileSystems, or whether it is a different list of FileSystems for different types of files, it can all be wrapped with a externally provided Env class, is it not?

mrambacher commented:
@dhruba I agree that the Env provides a nice container for abstracting everything OS out and is a nice container for the FileSystem objects (obviously the originators thought so too, as that is where it lives currently).  

As far as a stack or list of FileSystems inside of Env, both work fine.  The hard part is going to be validating that the set of FileSystems is valid for an existing DB.  But that will probably be true in both the stack and list cases (I do not believe there is any validation currently on the Env for a DB).

siying commented:
@dhruba what would be the interface in your proposal? Let's say, a RocksDB instance wants WAL and Level-0 files on local file system, and other levels on HDFS. What do you think the RocksDB options should look like for this setting?

dhruba commented:
Create a new env called CombineEnv. It will inherit from Env.  Its constructor
`
CombineEnv(HdfsConf configs) {
  // CombineEnv would have code to translate a local db path to its equivalent hdfs path. 
}

CombineEnv::NewWritableFile(CombineEnvOptions envoptions, file_level) {
  //The envoptions specified and file_level will be used to determine 
  // whether to create the file in HDFS or on local disk. If it is to be created in
  // Hdfs, then it will use HdfsConfig to map it to the right location.
}
`
Does that work?

siying commented:
@dhruba this interface is breaking software layering: **file_level**, a DB concept is exposed to Env layer, which is supposed to be a thin layer that abstracts out platform specific logic. This makes a future developer who creates Env to their environment very confusing.

dhruba commented:
I see your concern and agree with it. Can you specify the API-change you are proposing then in terms of some pseudo code prototype?

mrambacher commented:
@dhruba @siying This is a little outside of my area of RocksDB code, so perhaps this is a bit of a naive view/implementation.
- In Env, there is a map of names to FileSystems.  
- In DBOptions, we define a map of FileType to FileSystem name.  There is an array of these -- one for each level.  Additionally, there is a  FileType "default" entry in this array.
- If there is no entry in the map for a given type, the "default" FileSystem is used

When RocksDB goes to open a file, it sees what level and type it is and what level it is.  We look for the type in the Type/Name map for that level.  If we do not find it, we look for the "default" for that level.  If there is no default, we use the "default" file system.

Basically, this code would determine the proper FileSystem (by name/ID) for a given FileType by configuration.  Once it is found, we turn around and call the appropriate method of the FileSystem object.  There is nothing specific about Env for FileSystem, just a "Env::GetFileSystem(const std::string & id)" method.

siying commented:
@mrambacher what's the proposal really different from what we have now? `Env::GetFileSystem(const std::string & id)` can be added. I have no problem with that. It's not contradicting with this new API at all, and can make class Env conceptually one entity that abstracts out all platform related implementation.

The only different would be that DBOptions should take one or more of std::string or, a list of FileSystem objects directly. I prefer the latter interface because it avoids the other level of indirection from a critical path, so users don't have to worry about it unless they want. If users want to configure by string, they just call the function `Env::GetFileSystem(const std::string & id)` to get the translation.

Does it ease your concern?

mrambacher commented:
@siying I am not sure which "API" you are referring to here.  There are a few separate issues.  One is how one finds the FileSystem to use.  The code I see has a variable in db_impl that stores the FileSystem.  I do not believe it should be there but instead believe that the FileSystem should be a property/method of the Env class.  I believe the code should get a FileSystem ONLY through its Env and not store it separately.

As far as the API into the FileSystem itself, I still have concerns about having many methods take both an EnvOptions and an IOOptions.  And I do not like how the IODebugContext is used (I believe it should be a std::shared_ptr passed to the NewXXFile method and not to *every* method off the created file.

For configuration, there needs to be a way of serializing the FileSystem as part of the Env and reading it back into the DB.  Access via strings makes that portion easier, though I can see the advantage of adding FileSystems to Env directly as well.

siying commented:
@mrambacher it's a good point that we should keep string configuration in mind. String configuration should still work and for long term include Env and FileSystem too. @anand1976 do you already have any thoughts on it?

I don't see problem of DBImpl to have FileSystem inside. It's just like a lot of internal data structures have RandomAccessFile inside (which is a file descriptor wrapper). Just think about that a database holds some file descriptors of root of file system partitions. There is nothing wrong with databases to hold them inside. I don't see anything wrong with passing in those file descriptors to databases either.

I agree that it will be ideal that FileSystem doesn't have any reference to EnvOptions. I made the same comment already. I hope they are removed.

mrambacher commented:
@siying @anand1976 I have/had the Env code working for serialization.  I am willing to help with the same for FileSystem once it is complete.  However, it would be good to re-write *all* of the "Env FileSystems (HDFS, Posix, Encrypted, Mock) to use the new FileSystem module at the same time.  Otherwise, configurations will change and management will get more difficult.

I see no point in having a DBImpl store a FileSystem. How many does it store?  What is it storing that is not in the Env?  Note that a DBImpl is not something that is directly configured, so the FileSystem must be configured somewhere else.  And having a DBImpl store (and create) a FileSystem as a raw pointer seems dangerous -- what is the lifetime of that pointer.  I see no advantage to storing it and only problems related to it.

FileSystems are not on the same usage level as a File Descriptor, which I agree a DB needs to manage.

siying commented:
@mrambacher right now, in this implementation it only stores one file system per DB. In the future, it's likely that we would allow 2-3: one for WAL, one for hot SST files, and one for cold SST files. Maybe different column family can be different though. But either way, they will be pre-determined when DB is opened, or column family is created. I don't see a need to dynamically add a new file system during normal DB run.

mrambacher commented:
@siying I agree that dynamically adding a FileSystem to a DB is not something that is likely to be done.  The FileSystem(s) will be added to the DBOptions/Env prior to the DB starting.  But if the DBOptions goes out of scope, then storing a raw pointer to something in it would be bad.  And if the Env already has a pointer to the FileSystem, I question what the gain is to store the FileSystem again in the DBImpl (when it already has the Env).  It just creates an extra layer that has to be managed (if the Env changes from a FileSystem singleton to a map, then the DBImpl would also have to change.  More code to manage for no real benefit).

siying commented:
@mrambacher the pattern RocksDB is used is that, a copy will be made as DBImpl::initial_db_options_. With FileSystem stored as shared pointer there, it should be alive. This is how we make sure several other plug in function to be alive too, e.g., rater limiter, sst file manager, info logger, etc.

anand1976 commented:
> I have/had the Env code working for serialization. I am willing to help with the same for FileSystem once it is complete. However, it would be good to re-write all of the "Env FileSystems (HDFS, Posix, Encrypted, Mock) to use the new FileSystem module at the same time. Otherwise, configurations will change and management will get more difficult.

I haven't thought about serialization much. If you are able to help that would be great. As far as re-writing all the other file systems, I'd prefer to tackle it separately rather than this PR. But I agree that they should be ported to this new API as well.


> I see no point in having a DBImpl store a FileSystem. How many does it store? What is it storing that is not in the Env? Note that a DBImpl is not something that is directly configured, so the FileSystem must be configured somewhere else. And having a DBImpl store (and create) a FileSystem as a raw pointer seems dangerous -- what is the lifetime of that pointer. I see no advantage to storing it and only problems related to it.

Its actually stored in ImmutableDBOptions and ImmutableCFOptions. And its stored as a shared_ptr.

anand1976 commented:
@mrambacher 

> I am not sure which "API" you are referring to here. There are a few separate issues. One is how one finds the FileSystem to use. The code I see has a variable in db_impl that stores the FileSystem. I do not believe it should be there but instead believe that the FileSystem should be a property/method of the Env class. I believe the code should get a FileSystem ONLY through its Env and not store it separately.

IIUC, your proposal is to not have users specify which FileSystem to use in rocksdb::Options, but rather directly in the Env. Is that correct? And RocksDB queries the Env to get the appropriate FileSystem pointer. If that's the case, who makes the decision what filesystem to use for what purpose? Is it RocksDB's decision or the Env's decision?

anand1976 commented on env/fs_posix.cc:
```diff
@@ -0,0 +1,919 @@
+  IOStatus GetChildren(const std::string& dir, const IOOptions& /*opts*/,
+                       std::vector<std::string>* result,
+                       IODebugContext* /*dbg*/) override {
+    result->clear();
```
Specifying something when the File object is created would not work for retrieving fine grained statistics for a single operation. The IODebugContext is somewhat similar in concept to the IOStatsContext, which records some statistics for each IO call, but intended to provide more visibility into the Env which IOStatsContext does not. And unlike IOStatsContext, we want to move away from using thread local storage and explicitly pass a pointer to the object.

anand1976 commented on logging/posix_logger.h:
```diff
@@ -38,7 +38,10 @@ class PosixLogger : public Logger {
+      IOStatus status = IOError("Unable to close log file", "", ret);
+      if (!status.ok()) {
+        return Status::IOError("Unable to close log file", "");
+      }
```
Reverted this change. The return from IOError() can be directly returned.

siying commented on db/db_impl/db_impl_open.cc:
```diff
@@ -34,6 +35,14 @@ Options SanitizeOptions(const std::string& dbname, const Options& src) {
 DBOptions SanitizeOptions(const std::string& dbname, const DBOptions& src) {
   DBOptions result(src);
 
+  if (result.file_system == nullptr) {
```
Nothing wrong but I feel a little bit odd seeing comparing s shared_ptr to nullptr directly. Directly using `if (result.file_system)` feels more natural to me.

riversand963 replied:
I think it is OK because both `==` and `!` are overloaded: https://en.cppreference.com/w/cpp/memory/shared_ptr/operator_cmp

siying commented on include/rocksdb/options.h:
```diff
@@ -392,6 +393,8 @@ struct DBOptions {
   // Default: Env::Default()
   Env* env = Env::Default();
 
+  std::shared_ptr<FileSystem> file_system = nullptr;
```
Add a comment to explain it?

siying commented on options/db_options.cc:
```diff
@@ -101,6 +103,8 @@ void ImmutableDBOptions::Dump(Logger* log) const {
                    paranoid_checks);
   ROCKS_LOG_HEADER(log, "                                    Options.env: %p",
                    env);
+  ROCKS_LOG_HEADER(log, "                                     Options.fs: %p",
```
I think printing Name() is a better idea than the pointer.

siying commented on include/rocksdb/file_system.h:
```diff
@@ -0,0 +1,1345 @@
+// of the APIs is of type IOStatus, which can indicate an error code/sub-code,
+// as well as metadata about the error such as its scope and whether its
+// retryable.
+class FileSystem {
```
Can we add a Name() function, just like most configuration classes?

riversand963 commented on db/db_impl/db_impl_open.cc:
```diff
@@ -1295,10 +1310,10 @@ Status DBImpl::CreateWAL(uint64_t log_file_num, uint64_t recycle_log_number,
-    s = env_->ReuseWritableFile(log_fname, old_log_fname, &lfile,
-                                opt_env_options);
+    s = fs_->ReuseWritableFile(log_fname, old_log_fname, opt_file_options,
+                               &lfile, nullptr);
```
Nit: maybe consider add a short comment before `nullptr` like `/*var_name=*/nullptr `

riversand963 commented on db/error_handler.cc:
```diff
@@ -16,6 +16,22 @@ namespace rocksdb {
 // a specific error code or subcode.
+std::map<std::tuple<BackgroundErrorReason, bool>, Status::Severity>
+    RetryableErrorMap = {
+        {std::make_tuple(BackgroundErrorReason::kCompaction, true),
```
Annotating `true` with `/*retryable*/` can make it clearer. Or add a sentence to explain what it means.

anand1976 replied:
I'm actually planning to revert the changes to error_handler.cc. Originally, I combined the new API with some handling of retryable errors, but I think its better to separate it into another PR with a proper design.

riversand963 commented on db/wal_manager.cc:
```diff
@@ -464,9 +464,10 @@ Status WalManager::ReadFirstLine(const std::string& fname,
-      fname, &file, env_->OptimizeForLogRead(env_options_));
+  std::unique_ptr<FSSequentialFile> file;
+  Status status = fs_->NewSequentialFile(fname,
+                                         fs_->OptimizeForLogRead(file_options_),
```
Just curious, is this done by `make format`?

riversand963 commented on env/composite_env_wrapper.h:
```diff
@@ -0,0 +1,1230 @@
+
+namespace rocksdb {
+
+// The CompositeEnvWrapper class provides a interface that is compatible
```
an interface

riversand963 commented on env/composite_env_wrapper.h:
```diff
@@ -0,0 +1,1230 @@
+// Env or an Env + FileSystem. In the former case, the user will specify
+// Options::env only, whereas in the latter case, the user will specify
+// Options::env and Options::file_system.
+#define IO_STATUS_TO_STATUS(_status_, _io_status_)                 \
```
Can we inline this and avoid using macro?

anand1976 replied:
This macro can be removed. The wrapper functions can directly return the status, since Status is now the base class of IOStatus.

riversand963 commented on include/rocksdb/file_system.h:
```diff
@@ -0,0 +1,1345 @@
+
+ private:
+  // No copying allowed
+  FileSystem(const FileSystem&);
```
Nit: use `=delete`

riversand963 commented on include/rocksdb/options.h:
```diff
@@ -392,6 +393,8 @@ struct DBOptions {
   // Default: Env::Default()
   Env* env = Env::Default();
 
+  std::shared_ptr<FileSystem> file_system = nullptr;
```
`=nullptr` isn't necessary, IIRC

anand1976 replied:
Right, because its a shared_ptr and not raw.

anand1976 replied:
Although all other shared_ptr options in options.h initializes using ```= nullptr```, so maybe just leave it as is for consistency.

riversand963 commented on include/rocksdb/sst_file_manager.h:
```diff
@@ -110,6 +112,15 @@ class SstFileManager {
 //    directly recover them without checking.
+extern SstFileManager* NewSstFileManager(
+    Env* env, std::shared_ptr<FileSystem> fs,
+    std::shared_ptr<Logger> info_log = nullptr, std::string trash_dir = "",
```
const std::string& trash_dir = "",

riversand963 commented on env/composite_env_wrapper.h:
```diff
@@ -0,0 +1,1230 @@
+
+#pragma once
+
+#include <rocksdb/env.h>
```
I think we mostly use `#include "rocksdb/env.h"`. The difference is where to start the search. This is OK, but maybe it's good idea to be consistent across the codebase.

riversand963 commented on env/composite_env_wrapper.h:
```diff
@@ -0,0 +1,1230 @@
+class CompositeEnvWrapper : public Env {
+ public:
+  // Initialize an EnvWrapper that delegates all calls to *t
+  explicit CompositeEnvWrapper(Env* env, FileSystem* fs_env)
```
I find the name `fs_env` a little confusing. Can we rename it to `fs`?

riversand963 commented on env/composite_env_wrapper.h:
```diff
@@ -0,0 +1,1230 @@
+
+class CompositeEnvWrapper : public Env {
+ public:
+  // Initialize an EnvWrapper that delegates all calls to *t
```
We already have an `EnvWrapper` class. Does the `EnvWrapper` here refer to that?

anand1976 replied:
Sorry, it should be CompositeEnvWrapper.

riversand963 commented on db/db_impl/db_impl_open.cc:
```diff
@@ -233,6 +242,10 @@ Status DBImpl::ValidateOptions(const DBOptions& db_options) {
         "atomic_flush is incompatible with enable_pipelined_write");
   }
 
+  if (db_options.file_system != nullptr && !db_options.env) {
```
The default value of `options.env` is `Env::Default()`. Therefore, only when the application explicitly sets it to null can this happen. Is this correct?
In this case, will it work if we set it to `Env::Default()`?

anand1976 replied:
Yes, we can do that.

riversand963 commented on env/composite_env_wrapper.h:
```diff
@@ -0,0 +1,1230 @@
+class CompositeRandomAccessFileWrapper : public RandomAccessFile {
+ public:
+  explicit CompositeRandomAccessFileWrapper(
+      std::unique_ptr<FSRandomAccessFile>& target)
```
Should we use rvalue ref here?

anand1976 replied:
There are cases where we need to bind an lvalue to the argument, so rvalue ref won't work. For example, in ```CompositeEnvWrapper::NewRandomAccessFile()```.

riversand963 commented on env/fs_posix.cc:
```diff
@@ -0,0 +1,916 @@
+}
+
+#ifndef ROCKSDB_NO_DYNAMIC_EXTENSION
+class PosixDynamicLibrary : public DynamicLibrary {
```
I think `DynamicLibrary` should be part of Env, not FS.

riversand963 commented on env/env_posix.cc:
```diff
@@ -899,7 +272,8 @@ class PosixEnv : public Env {
     if (f == nullptr) {
       result->reset();
-      return IOError("when fopen a file for new logger", fname, errno);
+      // return IOError("when fopen a file for new logger", fname, errno);
```
Should we include error message, or clean up the comment?

anand1976 replied:
I'll change it. We should include the error message.

riversand963 commented on env/file_system.cc:
```diff
@@ -0,0 +1,108 @@
+  return s;
+}
+
+#ifdef OS_WIN
```
Should this be moved to the win directory?

anand1976 replied:
I added this here to keep the Windows build happy. When the Windows default env is ported to this new API, this would not be necessary anymore.

riversand963 commented on env/env_posix.cc:
```diff
@@ -1222,7 +525,9 @@ Env* Env::Default() {
   static PosixEnv default_env;
-  return &default_env;
+  static CompositeEnvWrapper composite_env(&default_env,
+                                           FileSystem::Default().get());
```
Is it by design that `FileSystem::Default()` always return the PosixFileSystem instance?

anand1976 replied:
Yes, along the lines of Env::Default().

anand1976 commented on include/rocksdb/file_system.h:
```diff
@@ -0,0 +1,1356 @@
+  virtual IOStatus NewSequentialFile(const std::string& fname,
+                                     const FileOptions& file_opts,
+                                     std::unique_ptr<FSSequentialFile>* result,
+                                     IODebugContext* dbg) = 0;
```
> * Why is IODebugConter rather than a std::shared_ptr<>?
> * Why do you need it at all? Why can't it be part of FileOptions?

I think I have addressed this before. The requirement is to be able to trace individual calls to the file system. The file system calls can be sampled, logged and then aggregated (or any other ad-hoc analysis) outside the IO path.

> * If it is not an output parameter, it should not be last

It is an output parameter.

mrambacher replied:
Seeing no examples of where this is being used, I still do not understand when you will trace one specific call for one specific file independent of other calls related to that file.  I can understand the use case of having a IODebugContext that is a parameter to the constructor of a file, but not saying that we are tracing only reads for a specific file.  I do not understand how the calling code would differentiate and pass a different DebugContext for file 1 versus file 2 for each of these calls.  

Is it an output parameter (as in something that is created) or an in-out parameter (as in something that might be changed -- more like a Logger class)?

anand1976 replied:
Its an in-out parameter. We follow the Google C++ style guide, which recommends placing input-only parameters before output parameters.

anand1976 commented on include/rocksdb/file_system.h:
```diff
@@ -0,0 +1,1356 @@
+};
+
+// A file abstraction for reading sequentially through a file
+class FSSequentialFile {
```
> Why can't the existing class be changed so that the old method (by default calls the new one? If you are worried about backwards compatibility, the new one can return "not implemented" and existing implementations will work. Alternatively, you have the new methods as pure virtual and legacy FileSystems return a FsFileWrapper class.

Having the new methods as pure virtual will break the compilation of derived class implementations. Our policy is to guarantee compatibility with any code written using public interfaces. 
The only way to make the approach of using existing classes is to define the new methods as not pure virtual and have them call the old methods by default. It might save some code change in other places, but those code changes are pretty straightforward and not risky, so why not bite the bullet now and design a clean set of interfaces?

> I think an approach similar to this also eliminates the need for the "Wrapper" classes to be handled at a higher layer. Now everything can implement the new interface coming out of the Env/FileSystem and the higher-level code does not need to deal with wrappers. As existing code is changed to use the new interfaces, it should just work

The FS* wrapper classes are not really referenced outside of composite_env_wrapper.h. The higher layers are not aware of them, except for some intrusive test code.

mrambacher replied:
I agree that it would be nice to have a clean break from the existing interface but do not believe it is practical.  To do so, you are required to implement an entirely new class and create all of these wrappers to it.  It would be so much simpler (and cleaner) to just have wrapper methods in the existing classes.    In the approach of adding new classes, you are effectively deprecating the old classes (instead of just deprecating the old methods), which creates two sets of classes that will be supported and maintained.  If you add the new methods, you can (eventually) declare the old ones to be deprecated.  I believe something similar to this probably happened in the past with a MergeOperator (and FullMerge vs FullMergeV2).

And the current approach is exposing wrappers all over the place (some to the File classes, some to the FileSystem).  A cleaner approach is to hide the "dirty laundry" of the wrapping inside the classes rather than exposing it outside.

anand1976 replied:
> And the current approach is exposing wrappers all over the place (some to the File classes, some to the FileSystem). A cleaner approach is to hide the "dirty laundry" of the wrapping inside the classes rather than exposing it outside.

I frankly don't understand this statement. The wrappers are not "exposed all over the place". There are a few non-critical places where the code has access to an Env and needs to wrap a FileSystem around it. That can very well be encapsulated into a function and the caller doesn't need to know anything about it. Does that address your concern?.

mrambacher replied:
Let me re-phrase:
1.  All of the existing code in Env dealing with FileSystem is left unchanged.  Which means there are now potentially two ways (the new and the old) to do everything.  And there is no comment or other indication to show that this code is potentially going away.
-> To solve this problem, you can make the Env have a FileSystem and wrap the existing methods into that FileSystem.

2.  Because the Env does not wrap the FileSystem, there is no way to test alternative FileSystem implementations.  Most of the tests assume that they should create a wrap of the FileSystem from the Env.
-> To solve this problem, see the first point.

3.  All of the existing code dealing with the Env FileXXX classes still remains, leading to two classes that implement very similar interfaces.
-> To solve this problem, have the "old" methods call the "new" methods in the class and add a comment that these methods will be deprecated in some future release.   Create a "LegacyFileWrapper" where the "new" method calls the old one in the wrapped target class.

4.  Because the new FSxxxFile classes are not the related to the old classes, many of the tests assume the old implementation and wrap it with the new one (LegacyWrapper).  This means that if we want to implement a Custom FileSystem, it must also be a part of an Env, meaning everything is there twice.

5.  There are many other Env in the base system (HDFS, Mock, ReportFileOp, ChRoot, etc) that should be changed to a "Basic Env with a FileSystem".  Have you tried to make any of these into FileSystem-based Environments?  I am wondering if the tests would pass under such conditions.

Also, there does not seem to be any mechanism to change what the FileSystem is in the DBOptions.  You cannot override it by an option in the INI file (there is no OptionTypeInfo defined for it).

anand1976 replied:
#1 and #3 are more about documentation and communication about deprecation of old interfaces, and is orthogonal to the design choice. I don't see having two classes, one old and one new, as a big problem. The old ones will be deprecated at some point.

I see your point about tests in #2 and #4. The CompositeEnvWrapper can be used to solve this, by returning the FileSystem that it wraps around. A default implementation in Env can return a LegacyFileSystemWrapper. Does that address your concern?

> Also, there does not seem to be any mechanism to change what the FileSystem is in the DBOptions.

Can we use FileSystem::Load() in ParseDBOptions(), similar to how Env is loaded?

mrambacher replied:
I think starting with an Env class that *has* a FileSystem in it and a "GetFileSystem" could solve many of the problems.  But I think you should do at least two implementations of the FileSystem to prove that it works.

If there is a FileSystem in an Env and in the DBOptions, how do you keep them in sync?  How do you know which one will be used for what operations?

anand1976 replied:
If the Env class has a FileSystem in it, how will a user specify multiple FileSystems if they have multiple DB instances in the same process? Do we have to allocate Env objects for the sole purpose of storing a FileSystem pointer? The default Env is a singleton. How can a user use the default Env and multiple FileSystems if the two are coupled together?

mrambacher commented on db/db_impl/db_impl.cc:
```diff
@@ -147,20 +149,21 @@ int64_t kDefaultLowPriThrottledRate = 2 * 1024 * 1024;
       own_info_log_(options.info_log == nullptr),
       initial_db_options_(SanitizeOptions(dbname, options)),
+      env_(initial_db_options_.env),
+      fs_(initial_db_options_.file_system),
```
Are you guaranteed that file_system is not NULL at this point?

anand1976 replied:
Yes, SanitizeOptions() would have allocated one if the user passed in NULL.

mrambacher commented on db/convenience.cc:
```diff
@@ -9,6 +9,7 @@
 #include "rocksdb/convenience.h"
 
 #include "db/db_impl/db_impl.h"
+#include "env/composite_env_wrapper.h"
```
Is this include still necessary?

anand1976 commented:
I'm going to land this PR. I had an offline discussion with @mrambacher. The summary is that we need to think about and address issues around making it easier for legacy Envs (such as HDFS, encrypted env etc.) to port to the new FileSystem+Env API. Some ideas discussed were storing the FileSystem pointer in the Env, and making the default Env not a singleton. Those need not block this PR and can be addressed in a separate PR.
